src/Emulator/Peripherals/Peripherals/Mocks/HPSHostController.cs

```diff
@@ -473,6 +473,7 @@ public void StartSocketServer(TimeInterval timeInterval)
                         if(data.Length == 1 && data[0] == 0x86)
                         {
                             PollForError(timeInterval);
+                            Thread.Sleep(10);
```

Comment:
To be frank I’m not sure if adding more `sleep` calls is the proper solution.

We should remember that the simulation in Renode is based on the virtual time, but the real time of execution depends heavily on the underlying hardware and the load. This means that using real-time `sleeps` verified to be enough on one platform might not function well in the future or on other platforms, e.g. in your CI.

That being said I would rather rework the solution to wait for particular events and use the virtual-time based timeouts. This would probably require implementing additional interfaces in the `HPSController` class itself. We can of course help you with that, but we would need to understand your specific expectation better.

Naturally we can discuss the details in another channel if you prefer.

Reply:
So the behaviour I observed back when I implemented the [PollForError](https://github.com/renode/renode-infrastructure/blob/63076c3ec0873ac5233999dce285e2ee7b0679b5/src/Emulator/Peripherals/Peripherals/Mocks/HPSHostController.cs#L369) was that, when issuing & reading data without sleep in between would end up returning `0,0` eventually (I assume could be an overload or the write is too fast), which is why I put a sleep statement after checking error each time. 

So after polling from the error register, if I issue the command immediately, it will return 0 (probably the same cause). But when I read the error register by `host.HPSHostController ReadError "1.0"`, the error bit is set for sure. I'm wondering what register bit could I check to make sure this won't happen. Tried PE, RXNE, TXIS, TCIE and TXIE.

Reply:
@mateusz-holenko I agree that inserting sleep ain't a long-term feasible solution, but we already have 100x longer sleep inserts inside PollForError.
Could you please provide more details, what the "particular event" would you wait for the virtual timeout?
I would like to scope the complexity of the solution you are proposing.

Reply:
My point is that we already wait for around 3 seconds of real time, so if adding another sleep of 10ms solves the problem it might just be a happy coincidence. In other words, you might need to add further sleeps when running on another system (like CI).

As for "particular event" - I might not understand the underlying logic of the binary simulated on STM32, but how do you know when the value of the `Error` register is ready to be read? Right now there is a polling for 3 real-time seconds, and if during this period it’s still set to all 0s this value is “accepted” as a correct one.
What if the simulation is scheduled differently and it would take 3.5s or 4s of real time for this register to be set?
I believe that this question is also valid on the hardware - at what point can you be sure that reading this register will return a correct value?

I see the following options - either it’s guaranteed to be set after a given amount of time (virtual time) or there is some flag (possibly in another register) indicating if the processing of the command is in progress or not. Polling of this value until it changes from 0 to 1 and then back to 0 could be a way of telling if the error register is ready to be read.

Making HPSController aware of the virtual time flow (implementing the ITimeSink interface and plugging it into the Time Framework) would allow for switching from real-time-based sleeps to virtual-time-based polling/waiting that in turn would allow for deterministic execution of the simulation, but this might require some work. 
Since you mentioned that merging the proposed change will unblock you, I’m fine with that but I consider this a workaround.

Reply:
Thanks for doing this, we definitely see this as a workaround too, but this would unblock us temporarily so that we may progress futher while debugging this issue. It would be great to get the change upstream, thanks you very much.